test/testutil.h

```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
  */
 
 /*
- * Returns 1 if |s1| and |s2| are both NULL or equal.
- * Otherwise, returns 0 and pretty-prints diagnostics using |desc|.
+ * A common routine to output test failure messages.  Generally this should not
+ * be called directly, rather it should be called by the following functions.
+ *
+ * |desc| is a printf formatted description with arguments |args|, |type| is
+ * the data type that was tested, |fmt| is a system provided printf format with
+ * arguments that spells out the failure details.
+ */
+#if defined(__GNUC__)
```

Comment:
The `fail_message` method signature is hard to comprehend. Can you give an example in the comments how it is supposed to be used, and what the resulting output looks like?

If this  is not meant to be used by tests (but only by the other macros), then perhaps move into a separate header?